src/library/scala/collection/immutable/Stream.scala

```diff
@@ -1293,6 +1244,23 @@ object Stream extends SeqFactory[Stream] {
   private[immutable] def collectedTail[A, B, That](head: B, stream: Stream[A], pf: PartialFunction[A, B], bf: CanBuildFrom[Stream[A], B, That]) = {
     cons(head, stream.tail.collect(pf)(bf).asInstanceOf[Stream[B]])
   }
-}
-
 
+  /** An implementation of `FilterMonadic` allowing GC of the `Stream` as it is processed.
+    *
+    * Because this is not an inner class of `Stream`, it is now possible to offer explicit
+    * 'null after first use' semantics for the reference to the head of the original
+    * `Stream`, allowing GC to collect while the tail is still processing (see SI-8990).
+    */
```

Comment:
I don't think we can make filtered Streams single-shot only. `Stream` is immutable, so I would expect the following to offer consistent results and not NPE.

```scala
% scala-hash v2.11.5
scala> val s = Stream(1, 2, 3, 4)
s: scala.collection.immutable.Stream[Int] = Stream(1, ?)

scala> val filtered = s withFilter (_ % 2 == 0)
filtered: s.StreamWithFilter = scala.collection.immutable.Stream$StreamWithFilter@1002d1c8

scala> filtered.map(identity).force
res7: scala.collection.immutable.Stream[Int] = Stream(2, 4)

scala> filtered.map(identity).force
res8: scala.collection.immutable.Stream[Int] = Stream(2, 4)
```